components/camel-google-bigquery/src/main/java/org/apache/camel/component/google/bigquery/GoogleBigQueryProducer.java

```diff
@@ -0,0 +1,219 @@
+        for (Exchange ex: exchanges) {
+            Object entryObject = ex.getIn().getBody();
+            if (entryObject instanceof List) {
+                for (Map<String, Object> entry: (List<Map<String, Object>>) entryObject) {
+                    apiRequestRows.add(createRowRequest(null, entry));
+                }
+            } else if (entryObject instanceof Map) {
+                apiRequestRows.add(createRowRequest(ex, (Map<String, Object>) entryObject));
+            } else {
+                ex.setException(new IllegalArgumentException("Cannot handle body type " + entryObject.getClass()));
+            }
+        }
+
+        if (apiRequestRows.isEmpty()) {
+            return 0;
+        }
+
+        GoogleBigQueryEndpoint endpoint = getEndpoint();
+
+        TableDataInsertAllRequest apiRequestData = new TableDataInsertAllRequest().setRows(apiRequestRows);
+
+        Bigquery.Tabledata.InsertAll apiRequest = endpoint.getBigquery()
+                .tabledata()
+                .insertAll(configuration.getProjectId(),
+                        configuration.getDatasetId(),
+                        tableIdWithPartition,
+                        apiRequestData);
+        if (suffix != null) {
+            apiRequest.set("template_suffix", suffix);
+        }
+
+        logger.trace("uploader thread/id: {} / {} . calling google api", Thread.currentThread(), exchangeId);
+
+        TableDataInsertAllResponse apiResponse = apiRequest.execute();
+
+        if (apiResponse.getInsertErrors() != null && !apiResponse.getInsertErrors().isEmpty()) {
+            throw new Exception("InsertAll into " + tableId + " failed: " + apiResponse.getInsertErrors());
+        }
+
+        logger.debug("uploader thread/id: {} / {} . api call completed.", Thread.currentThread().getId(), exchangeId);
+        return apiRequestData.size();
+    }
+
+    private TableDataInsertAllRequest.Rows createRowRequest(Exchange exchange, Map<String, Object> object) {
+        TableRow tableRow = new TableRow();
+        tableRow.putAll(object);
+        String insertId = null;
+        if (configuration.getUseAsInsertId() != null) {
+            insertId = (String)(object.get(configuration.getUseAsInsertId()));
+        } else {
+            if (exchange != null) {
+                insertId = exchange.getIn().getHeader(GoogleBigQueryConstants.INSERT_ID, String.class);
+            }
+        }
+        TableDataInsertAllRequest.Rows rows = new TableDataInsertAllRequest.Rows();
+        rows.setInsertId(insertId);
+        rows.setJson(tableRow);
+        return rows;
+    }
+
+    @Override
+    public boolean isSingleton() {
+        return true;
+    }
+
+    @Override
+    public GoogleBigQueryEndpoint getEndpoint() {
+        return (GoogleBigQueryEndpoint) super.getEndpoint();
+    }
+
+    @Override
+    public boolean process(Exchange exchange, AsyncCallback callback) {
+        getEndpoint().getExecutorService().submit(() -> {
```

Comment:
As the executor is only used here, maybe the Processor could take executor as constructor argument.